smpp/pdu/pdufield/tlvtypes.go

```diff
@@ -0,0 +1,586 @@
+	// parameter must occur an equal number of instances and the order of occurrence determines the particular
+	// callback_num_atag which corresponds to a particular callback_num
+	CallbackNumAtag TLVType
+
+	// SourceSubAddress The subaddress of the message originator.
+	SourceSubAddress TLVType
+
+	// DestSubAddress The subaddress of the message destination
+	DestSubAddress TLVType
+
+	// UserResponseCode A user response code. The actual response codes are implementation specific
+	UserResponseCode TLVType
+
+	// DisplayTime  Provides the receiving MS based SME with a display time associated with the message
+	DisplayTime TLVType
+
+	// SMSSignal  Indicates the alerting mechanism when the message is received by an MS
+	SMSSignal TLVType
+
+	// MSValidity  Indicates validity information for this message to the recipient MS
+	MSValidity TLVType
+
+	// MsMsgWaitFacilities  This parameter controls the indication and specifies the message
+	// type (of the message associated with the MWI) at the mobile station.
+	MsMsgWaitFacilities TLVType
+
+	// NumberOfMessages  Indicates the number of messages stored in a mail box (e.g. voice mail box)
+	NumberOfMessages TLVType
+
+	// AlertOnMsgDelivery Requests an MS alert signal be invoked on message delivery
+	AlertOnMsgDelivery TLVType
+
+	// LanguageIndicator  Indicates the language of an alphanumeric text message.
+	LanguageIndicator TLVType
+
+	// ItsReplyType  The MS user’s reply method to an SMS delivery message received from the network
+	// is indicated and controlled by this parameter
+	ItsReplyType TLVType
+
+	// ItsSessionInf Session control information for Interactive Teleservice
+	ItsSessionInfo TLVType
+}
+
+type dataSMRespTLVParameter struct {
+	// DeliveryFailureReason  Include to indicate reason for delivery failure
+	DeliveryFailureReason TLVType
+
+	// NetWorkErrorCode  Error code specific to a wireless network
+	NetWorkErrorCode TLVType
+
+	// AdditionalStatusInfoText ASCII text giving a description of the meaning of the response
+	AdditionalStatusInfoText TLVType
+
+	// DPFResult Indicates whether the Delivery Pending Flag was set
+	DPFResult TLVType
+}
+
+// BindTransmiterTLVParameter defines the TLV parameters available  for BindTrasmiter PDU
+var BindTransmiterTLVParameter *bindTransmiterTLVParameter
+
+// SubmitSMTLVParameter defines the TLV parameters available for SubmitSM PDU
+var SubmitSMTLVParameter *submitSMTLVParameter
+
+// SubmitSMMultiTLVParameter defines the TLV parameters available for submitSMMulti PDU
+var SubmitSMMultiTLVParameter *submitSMMultiTLVParameter
+
+// DeliverSMTLVParameter defines TLV parameters available for DeliverSM PDU
+var DeliverSMTLVParameter *deliverSMTLVParameter
+
+//DataSMTLVParameter defines the TLV parameters available for DataSM PDU
+var DataSMTLVParameter *dataSMTLVParameter
+
+// DataSMRespTLVParameter defines the TLV parameters available for DataSMResp PDU
+var DataSMRespTLVParameter *dataSMRespTLVParameter
```

Comment:
Please change each PDU type allowed TLV fields to a map, they shouldn't be structs.